streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```
@@ -368,6 +370,7 @@ public void setUncaughtExceptionHandler(final Thread.UncaughtExceptionHandler eh
      * @param queryableStoreType    accept only stores that are accepted by {@link QueryableStoreType#accepts(StateStore)}
      * @param <T>                   return type
      * @return  A facade wrapping the {@link org.apache.kafka.streams.processor.StateStore} instances
+     * @throws org.apache.kafka.streams.errors.InvalidStateStoreException if the streams are (re-)initializing
```

A few correlated comments:

1. Seems we are doing redundant check of `allStores.isEmpty()` in `QueryableStoreProvider` as the same check is also conducted in the returned `WrappingStoreProvider`.

2. And in `WrappingStoreProvider`, we could just return the empty list if `allStores.isEmpty()`; and similarly in `StreamThreadStateStoreProvider`, simply remove the `if (!store.isOpen())` check and still put it into the list since the check will be conducted later on the `get` / etc calls as well.

3. Finally in the `CompositeReadOnlyKeyValueStore` and `CompositeReadOnlyWindowStore`, check if `stores.isEmpty()` and also catch-rethrow `InvalidStateStoreException` thrown from `store.get/range/etc` calls to re-write the error message from

```
Store X is currently closed.
```

to

```
State store is not available anymore and may have been migrated to another instance; please re-discover its location from the state metadata.
```

And if `stores.isEmpty` also throw the `InvalidStateStoreException` with the same error message.

My understanding is that we can be ensured this is the case since in `onPartitionsAssigned`, we always update the state metadata after all the tasks have been added (i.e. stores being constructed and opened), is that right?

---

And then for this `store` function, we can remove the possible thrown `InvalidStateStoreException` as well.

---

@guozhangwang 

1. I don't think the check is redundant. In `QueryableStoreProvider` this is called from `KafkaStreams.store(...)` there may not exist any stores for the `name` and `type` in the streams instance - we should throw an exception here as we can't currently find the store. 

2.  We probably don't need empty check in `WrappingStoreProvider` as an exception will bubble up from `StreamThreadStateStoreProvider` if there is a rebalance.

3. I don't think we need any empty check in `CompositeReadOnlyKeyValueStore` and `CompositeReadOnlyWindowStore` as the `InvalidStateStoreException` from `StreamThreadStoreProvider` with the appropriate message will bubble up. 

Yes we update the state metadata after all the tasks and stores have been initialized

---

Actually, thinking about it a bit more i think that for point 2 we do actually want to do the empty check and throw `InvalidStateStoreException`. Why? if there was a rebalance and the store migrated elsewhere, but this call happens after the migration and `allStores.isEmpty()` then this indicates that the store is no longer on this instance, so we should throw.
i.e,:
```
store = streams.store(...); // store exists at this point and not rebalancing
// some time passes and a rebalance happen, store gets migrated elsewhere
store.get("blah");  // should throw InvalidStateStoreException as the store no longer exists in this instance